src/validationinterface.cpp

```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
     boost::signals2::signal<void (const CBlockIndex *, const std::shared_ptr<const CBlock>&)> NewPoWValidBlock;
 
     CScheduler *scheduler = NULL;
+
+    // We are not allowed to assume the scheduler only runs in one thread,
+    // but must ensure all callbacks happen in-order, so we end up creating
+    // our own queue here :(
+    CCriticalSection cs_callbacksPending;
+    std::list<std::function<void (void)>> callbacksPending;
+    bool fCallbacksRunning = false;
+
+    void MaybeScheduleProcessQueue() {
+        {
+            LOCK(cs_callbacksPending);
+            // Try to avoid scheduling too many copies here, but if we
```

Comment:
@ryanofsky I see that now, the extra check does prevent the execution. 

Being that this is new code I wouldn't call it a simplification. Here's a patch of the proposed change, less logic with the same function:
[scheduler.patch.txt](https://github.com/bitcoin/bitcoin/files/977512/scheduler.patch.txt)

which reads better if you rename `fCallbacksRunning` to `fCallbacksScheduled`
and this patch:
[scheduler.patch2.txt](https://github.com/bitcoin/bitcoin/files/977548/scheduler.patch2.txt)
which can be argued reduces code reuse but I think the readability is improved.